src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    //TODO: use fine-grained lock if possible
+    public synchronized boolean takeSnapshot() {
+        logger.info("Start generating the snapshot");
+
+        if (this.isTombstoneFilesMerging) {
+            logger.error("DB is not ready for snapshot now");
+            return false;
+        }
+
+        try {
+            final int currentWriteFileId;
+            compactionManager.pauseCompactionThread();
+
+            // Only support one snapshot now
+            // TODO: support multiple snapshots if needed
+            File snapshotDir = getSnapshotDirectory();
+            if (snapshotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapshotDir);
+            }
+
+            FileUtils.createDirectoryIfNotExists(snapshotDir);
+            logger.info("Created directory for snapshot {}", snapshotDir.toString());
+
+            writeLock.lock();
+            // TODO: Need solutions to handle the file pieces after the rollover
```

Comment:
This may not be needed as long as the percentage of stale data is well tracked. I would suggest to remove this comment. Or it may give user an impression this is a half done work. As long as it works well, TODO == WILL NEVER DO

Reply:
In fact, the data and datac file would be well maintained. The issue is the tombstone file, it would not be compacted till the restart of the storage engine. That maybe a issue for long run. Looks related to existing issue which the tombstone can't be compacted until the restart.